### PR TITLE
Handle unexpected keyword rather than identifier for lambda parameter name

### LINKED ISSUE #60661: Lambda using 'sbyte' as parameter name overflows the parsing stack
_This issue has been moved from [a ticket on Developer Community](https://developercommunity.visualstudio.com/t/Code-breaks-syntax-coloring/10007337)._

Minimal repro is just `x.Where(sbyte =>)` by itself in a source file. [SharpLab](https://sharplab.io/#v2:EYLgtghglgdgNAExAagD4A8B0B1AFgUwCd8AKAZ2AE8AXfAAgF4A+ASgFgAoIA==)

---
[severity:It's more difficult to complete my work]

```cs
using Microsoft.EntityFrameworkCore;

using System.Diagnostics;

using Tempus.Core.TreeStructures;
using Tempus.Core.TreeStructures.Helpers;
using Tempus.Data.Entities;
using Tempus.Data.Models;

using static Tempus.Core.Mappers.IMapper;

namespace Tempus.Data
{
    public class CachedStavebniObjektDAO : DaoBase, IStavebniObjektDAO
    {
        private readonly IMapper mapper;

        public CachedStavebniObjektDAO(IDbContextFactory contextFactory,
            IMapper mapper) : base(contextFactory)
        {
            this.mapper = mapper;
        }

        private readonly Dictionary Cache = new Dictionary();

        public void DBWarmUpCall()
        {
            using (var db = DB)
            {
                var query = from user in db.USER
                            select 1;
                _ = query.FirstOrDefault();
            }
        }

        /// <summary>
        /// Loads the cache from the database.
        /// <br>Typically this is to be called once when the server starts 
        /// and then upon every change in the table, the cache needs to be refreshed</br>
        /// </summary>
        public void LoadAll()
        {
            Cache.Clear();
            var items = GetAllFlatStructure();

            Stopwatch s = Stopwatch.StartNew();

            //put in cache
            foreach(var item in items)
            {
                Cache.Add(item.ID, item);
            }

            TreeEntityHelpers.BuildTree(Cache);

            foreach(var item in items)
            {
                item.FillPathToRoot();
                item.FillPathFromRoot();
            }
            s.Stop();
            Console.WriteLine($"tree operations took {s.ElapsedMilliseconds} ms");
        }


        public List GetAllFlatStructure()
        {
            using (var db = DB)
            {
                var view = db.STAVEBNI_OBJEKT
                    .Include(s => s.Projekt).ThenInclude(p => p.KONFIGURACE_PROJEKTU).ThenInclude(k => k.SadaUrovniSO)
                    .Include(s => s.Projekt).ThenInclude(p => p.KONFIGURACE_PROJEKTU).ThenInclude(k => k.TypKonfigurace);

                var query = from s in view
                            select mapper.Map(s);
                var items = query.AsNoTracking().ToList();
                return items;
            }
        }

        /// <summary>
        /// 
        /// </summary>
        /// If not present
        private StavebniObjektEntity FindInCache(int ID)
        {
            return Cache[ID];
        }

        public StavebniObjektEntity Get(int ID)
        {
            return FindInCache(ID);
        }

        public List GetAll()
        {
            return Cache.Values.ToList();
        }

        public List GetAll(int projektID)
        {
            return GetAll.Where(sbyte =>)
        }
    }
}

```

---
#### Fiona Niu[MSFT] on 4/7/2022, 08:24 AM: 

(private comment, text removed)
#### Jan Seriš on 4/7/2022, 05:28 PM: 

(private comment, text removed)
#### Feedback Bot on 4/8/2022, 06:38 AM: 

(private comment, text removed)
#### Feedback Bot on 4/8/2022, 09:21 PM: 

(private comment, text removed)

---
### Original Solutions
(no solutions)

### PR DESCRIPTION
**Pull request by cston**
Fixes #60661

### CODE REVIEW

File: src/Compilers/CSharp/Test/Semantic/Semantics/LambdaTests.cs [hunk 1]
```diff
@@ -6310,6 +6310,26 @@ class A : Attribute { }
...
+        [WorkItem(60661, "https://github.com/dotnet/roslyn/issues/60661")]
+        public void KeywordParameterName()
+        {
+            var source =
+@"using System;
+class Program
+{
+    static void Main()
+    {
+        Action<int> a = int => { };
```
Comment (Youssef1313) [inline src/Compilers/CSharp/Test/Semantic/Semantics/LambdaTests.cs:6323]: Consider adding another test for `public`. Currently the behavior for `int` (stack exhausting) is different than `public` (too many syntax errors).

Also consider testing with `@int` (not too important)

Comment (cston) [reply src/Compilers/CSharp/Test/Semantic/Semantics/LambdaTests.cs:6323]: Added a parsing test for a full syntax tree with `public`.

**Comment by cston** [comment]
@dotnet/roslyn-compiler, please review, thanks.

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 2]
```diff
@@ -13076,7 +13076,9 @@ LambdaExpressionSyntax parseLambdaExpressionWorker()
                     // Unparenthesized lambda case
                     // x => ...
                     // x!! => ...
-                    var identifier = this.ParseIdentifierToken();
+                    var identifier = (this.CurrentToken.Kind != SyntaxKind.IdentifierToken && this.PeekToken(1).Kind == SyntaxKind.EqualsGreaterThanToken) ?
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:13079]: Could you clarify why we need to peek for an `=>` here? It seems like we already scanned and determined that this is a lambda?

Comment (cston) [reply src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:13079]: At this point, we may have `x => ...` or just `=> ...` (missing parameter name).

**Comment by cston** [comment]
@dotnet/roslyn-compiler, please review, thanks.

File: src/Compilers/CSharp/Test/Semantic/Semantics/LambdaTests.cs (hunk 1 above)
Comment (333fred) [inline src/Compilers/CSharp/Test/Semantic/Semantics/LambdaTests.cs:6323]: Could you add a test for `ref => {}`?

Comment (333fred) [reply src/Compilers/CSharp/Test/Semantic/Semantics/LambdaTests.cs:6323]: And `ref int => {}`

**Review by jcouv — APPROVED** [review]
LGTM Thanks (iteration 4)

**Review by 333fred — APPROVED** [review]
LGTM (commit 6)

File: src/Compilers/CSharp/Test/Syntax/Parsing/LambdaParameterParsingTests.cs [hunk 3]
```diff
@@ -3035,5 +3035,445 @@ public void AsyncAwaitInLambda()
...
             EOF();
         }
+
+        [WorkItem(60661, "https://github.com/dotnet/roslyn/issues/60661")]
+        [InlineData(LanguageVersion.CSharp9)]
+        [InlineData(LanguageVersionFacts.CSharpNext)]
+        [Theory]
+        public void KeywordParameterName_01(LanguageVersion languageVersion)
+        {
+            string source = "int =>";
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Test/Syntax/Parsing/LambdaParameterParsingTests.cs:3045]: Is this only an issue for non-parenthesized lambdas? What if we had `(int, int) =>`?

Comment (RikkiGibson) [reply src/Compilers/CSharp/Test/Syntax/Parsing/LambdaParameterParsingTests.cs:3045]: It might also be good to just test `(int) =>` as well.

**Pull request merged**